### ISSUE TITLE
User Tracking Mode

### ISSUE THREAD

**Issue by d3mac123** [Yermo/nativescript-mapbox#144]
Is it possible to "follow" the user location, auto rotating the map? For example, if I am walking towards the south, I would like the map "following me", auto-rotating from N to S. 

Something like this setting (from the React Mapbox plugin):
```
userTrackingMode  ->Whether the map is zoomed to and follows the user's location. One of:
Mapbox.userTrackingMode.none, 
Mapbox.userTrackingMode.follow, 
Mapbox.userTrackingMode.followWithCourse, 
Mapbox.userTrackingMode.followWithHeading
```
I have tried checking the plugin code to see how difficult is to implement this. It does not seem a big deal but, unfortunately, as I never coded natively for iOS or Android, I have no idea what to do there :)

**Comment by b4rtt** [comment]
I tried to find way how to do it too. But without success...

**Comment by d3mac123** [comment]
Eddy (and other wizards of native code),

If it helps, this is the documentation from Mapbox site:
https://www.mapbox.com/ios-sdk/api/3.3.4/Enums/MGLUserTrackingMode.html#/c:@E@MGLUserTrackingMode@MGLUserTrackingModeFollow

**Comment by d3mac123** [comment]
Hey @EddyVerbruggen, any chance you help me here? I know I am pretty close to making the map automatically track the user heading/course.

I've added the following lines into your mapbox.ios.js file (under the MapboxView.prototype.initMap function):
`console.log("User tracking mode: "+MGLUserTrackingMode+", "+MGLMapView.userTrackingMode+", "+MGLUserTrackingModeFollowWithCourse);
MGLUserTrackingMode = MGLUserTrackingModeFollowWithCourse`

The console.log shows the number 3 for the MGLUserTrackingModeFollowWithCourse (with is the right number according to Mapbox's API documentation). However, I can't find a way to make it work with the `MGLUserTrackingMode = MGLUserTrackingModeFollowWithCourse` code. Do you know the right way to set it?

Thanks a lot,
Alex

**Comment by d3mac123** [comment]
I finally got how to set the userTrackingMode to work on iOS (this is an ugly implementation, forced manually into @EddyVerbruggen code at mapbox.ios.js).

I have chosen to add this line in the setCenter function (at it constantly runs when location changes):
```
Mapbox.prototype.setCenter = function (options, nativeMap) {
        return new Promise(function (resolve, reject) {
            try {
                var theMap = nativeMap || _mapbox.mapView;
                var animated = options.animated === undefined || options.animated;
                theMap.userTrackingMode = MGLUserTrackingMode.MGLUserTrackingModeFollowWithCourse;
                var coordinate = CLLocationCoordinate2DMake(options.lat, options.lng);
                theMap.setCenterCoordinateAnimated(coordinate, animated);
                resolve();
            }
            catch (ex) {
                console.log("Error in mapbox.setCenter: " + ex);
                reject(ex);
            }
        });
};
```

The line I have added is this one:
`theMap.userTrackingMode = MGLUserTrackingMode.MGLUserTrackingModeFollowWithCourse;`

Hopefully, someone with better skills will add this as a map setting option and commit it in a public release (I have no idea how to do that).

Now, trying to do the same with Android.

**Comment by d3mac123** [comment]
Here's the solution for Android (like above, add this code in the setCenter function (at it constantly runs when location changes):
```
Mapbox.prototype.setCenter = function (options, nativeMap) {
        return new Promise(function (resolve, reject) {
            try {
                var theMap = nativeMap || _mapbox;
                theMap.mapboxMap.getTrackingSettings().setMyLocationTrackingMode(com.mapbox.mapboxsdk.constants.MyLocationTracking.TRACKING_FOLLOW);
                theMap.mapboxMap.getTrackingSettings().setMyBearingTrackingMode(com.mapbox.mapboxsdk.constants.MyBearingTracking.GPS);
                var cameraPosition = new com.mapbox.mapboxsdk.camera.CameraPosition.Builder()
                    .target(new com.mapbox.mapboxsdk.geometry.LatLng(options.lat, options.lng))
                    .build();
                if (options.animated === true) {
                    theMap.mapboxMap.animateCamera(com.mapbox.mapboxsdk.camera.CameraUpdateFactory.newCameraPosition(cameraPosition), 1000, null);
                }
                else {
                    theMap.mapboxMap.setCameraPosition(cameraPosition);
                }
                resolve();
            }
            catch (ex) {
                console.log("Error in mapbox.setCenter: " + ex);
                reject(ex);
            }
        });
    };
```

The lines I have added are:
```
theMap.mapboxMap.getTrackingSettings().setMyLocationTrackingMode(com.mapbox.mapboxsdk.constants.MyLocationTracking.TRACKING_FOLLOW);
theMap.mapboxMap.getTrackingSettings().setMyBearingTrackingMode(com.mapbox.mapboxsdk.constants.MyBearingTracking.GPS);
```

**Comment by EddyVerbruggen** [comment]
Let me play with this (soon). Not sure this needs to be configurable instead of changing the implementation.

**Comment by EddyVerbruggen** [comment]
I'm about to release 4.1.0 which will get a new `trackUser` function to enable this feature. Docs and demo have been updated.

**Comment by d3mac123** [comment]
You're, honestly, an amazing professional. Thank you very much for making Nativescript easier.

Alex
On Jul 2, 2018, 11:11 AM -0400, Eddy Verbruggen <notifications@github.com>, wrote:

Closed #144<https://github.com/EddyVerbruggen/nativescript-mapbox/issues/144>.
